src/csharp/Grpc.IntegrationTesting/ExternalDnsWithTracingClientServerTest.cs

```diff
@@ -0,0 +1,171 @@
+            client = new TestService.TestServiceClient(channel);
+        }
+
+        [OneTimeTearDown]
+        public void Cleanup()
+        {
+            channel.ShutdownAsync().Wait();
+            server.ShutdownAsync().Wait();
+        }
+
+        [Test]
+        public void EmptyUnary()
+        {
+            InteropClient.RunEmptyUnary(client);
+            Assert.Greater(newLogger.LogCount, 0);
+        }
+    }
+
+    /// <summary>
+    /// Logger which does some socket operation after delegating
+    /// actual logging to its delegate logger. The main goal is to
+    /// reset the current thread's WSA error status.
+    /// The only reason for the delegateLogger is to continue
+    /// to have this test display debug logs.
+    /// </summary>
+    internal sealed class SocketUsingLogger : ILogger
+    {
+        private ILogger delegateLogger;
+        private object myObject = new object();
+        private int logCount = 0;
+
+        public SocketUsingLogger(ILogger delegateLogger) {
+            this.delegateLogger = delegateLogger;
+        }
+
+        public void Debug(string message)
+        {
+            MyLog(() => delegateLogger.Debug(message));
+        }
+
+        public void Debug(string format, params object[] formatArgs)
+        {
+            MyLog(() => delegateLogger.Debug(format, formatArgs));
+        }
+
+        public void Error(string message)
+        {
+            MyLog(() => delegateLogger.Error(message));
+        }
+
+        public void Error(Exception exception, string message)
+        {
+            MyLog(() => delegateLogger.Error(exception, message));
+        }
+
+        public void Error(string format, params object[] formatArgs)
+        {
+            MyLog(() => delegateLogger.Error(format, formatArgs));
+        }
+
+        public ILogger ForType<T>()
+        {
+            return this;
+        }
+
+        public void Info(string message)
+        {
+            MyLog(() => delegateLogger.Info(message));
+        }
+
+        public void Info(string format, params object[] formatArgs)
+        {
+            MyLog(() => delegateLogger.Info(format, formatArgs));
+        }
+
+        public void Warning(string message)
+        {
+            MyLog(() => delegateLogger.Warning(message));
+        }
+
+        public void Warning(Exception exception, string message)
+        {
+            MyLog(() => delegateLogger.Warning(exception, message));
+        }
+
+        public void Warning(string format, params object[] formatArgs)
+        {
+            MyLog(() => delegateLogger.Warning(format, formatArgs));
+        }
+
+        public int LogCount {
+          get { return logCount; }
+        }
+
+        private void MyLog(Action delegateLog) {
+          lock (myObject) {
+              logCount++;
+          }
+          delegateLog();
+          Socket s = new Socket(AddressFamily.InterNetwork,
```

Comment:
add a comment what's the purpose of this?

Reply:
done